### DISCUSSION TITLE
B_SelectSpell.d

### DISCUSSION THREAD

**Discussion by Kor-Angar** [dosinabox/g2nr_unofficial_update#520] [General]
при текущем прохождению потестирую переписанный B_SelectSpell

**Comment by Kor-Angar** [comment]
```D
        // ------ Random fьr eine Serie ------
        if (slf.aivar[AIV_SelectSpell] <= 0)
        {
            dK_rnd = Hlp_Random(10);
            slf.aivar[AIV_SelectSpell] += dK_rnd; //+ 0-9
        };
```

тут было бы логичнее сделать 

```
slf.aivar[AIV_SelectSpell] = dK_rnd+1; // 1-10
```

собственно возможно из-за этого ищущие и тупят иногда

**Reply by dosinabox** [reply]
> slf.aivar[AIV_SelectSpell] = dK_rnd+1;

Сделай так у себя и протестируй в этом прохождении, пожалуйста

**Reply by Kor-Angar** [reply]
угу

**Reply by Kor-Angar** [reply]
не.. не так

**Reply by Kor-Angar** [reply]
if (slf.aivar[AIV_SelectSpell] < 0)
        {
            dK_rnd = Hlp_Random(10);
            //slf.aivar[AIV_SelectSpell] += dK_rnd; //+ 0-9
            slf.aivar[AIV_SelectSpell] = dK_rnd; // 0-9
        };

**Reply by Kor-Angar** [reply]
херня какая то
![image](https://github.com/dosinabox/g2nr_unofficial_update/assets/49514764/3c351ad1-2e0c-4319-aa96-352c198a1689)

@dosinabox  каким образом `AIV_SelectSpell` может принять отрицательное значение?

**Reply by Kor-Angar** [reply]
![image](https://github.com/dosinabox/g2nr_unofficial_update/assets/49514764/44bb2c4a-7bf7-4c4c-b489-6e060d3ef903)
возможно просто обнуление будет  достаточно перенести в начало

**Comment by Kor-Angar** [comment]
**Ищущие**

```D
 	// ------ Dementor -------
    if (slf.guild == GIL_DMT) 
    && (slf.aivar[AIV_MagicUser] == MAGIC_ALWAYS) // излишняя проверка?
    {
        /*
        // ------- in die Taschen -------
        if (Npc_HasItems (slf, ItRu_InstantFireball) == 0)	{	CreateInvItems (slf, ItRu_InstantFireball, 1);	};
        if (Npc_HasItems (slf, ItRu_Deathbolt) == 0) 		{	CreateInvItems (slf, ItRu_Deathbolt, 1); };
        if (Npc_HasItems (slf, ItRu_Deathball) == 0) 		{	CreateInvItems (slf, ItRu_Deathball, 1); };
        if (Npc_HasItems (slf, ItRu_Firerain) == 0) 		{	CreateInvItems (slf, ItRu_Firerain, 1); };
        if (Npc_HasItems (slf, ItRu_Thunderstorm) == 0) 	{	CreateInvItems (slf, ItRu_Thunderstorm, 1); };
        if (Npc_HasItems (slf, ItRu_LightningFlash) == 0) 	{	CreateInvItems (slf, ItRu_LightningFlash, 1); };
        if (Npc_HasItems (slf, ItRu_Firestorm) == 0) 		{	CreateInvItems (slf, ItRu_Firestorm, 1); };
        if (Npc_HasItems (slf, ItRu_Skull) == 0) 			{	CreateInvItems (slf, ItRu_Skull, 1); };
						
        // ------------------------------
        */

        if(C_IsNpc(slf,DMT_1299_OberDementor_DI))
        {
            //B_ReadySpell(slf,SPL_Skull,SPL_Cost_Skull);
            return  B_ReadyRune(slf,ItRu_Skull);
        };
        
        
        
        //alle anderen Dementoren 
        
        if(Npc_IsDrawingWeapon(slf))
        {
            return TRUE;	// nichts tun wenn gerade am waffe ziehen
        };
        
        Hlp_PrintConsole(Str_Format("  B_SelectSpell[%s] --> AIV_SelectSpell = %i", slf.name, slf.aivar[AIV_SelectSpell]));
        
        if(slf.aivar[AIV_SelectSpell] == 12) // neuer spell erst wenn der alte abgeschossen wurde
        {
            slf.aivar[AIV_SelectSpell] = 0;
        };  
        
        // ------ Random fьr eine Serie ------
        if (slf.aivar[AIV_SelectSpell] <= 0)
        {
            //dK_rnd = Hlp_Random(10);
            //slf.aivar[AIV_SelectSpell] += dK_rnd; //+ 0-9
            dK_rnd = Hlp_Random(9);
            slf.aivar[AIV_SelectSpell] = dK_rnd+1; // 1-9
            Hlp_PrintConsole(Str_Format("  B_SelectSpell[%s] --> Hlp_Random --> AIV_SelectSpell = %i", slf.name, slf.aivar[AIV_SelectSpell]));
        };
        
        if(slf.aivar[AIV_SelectSpell] < 10)
        {
            if (Kapitel <= 3) { return  B_ReadyRune(slf,ItRu_InstantFireball); };   // 75
            if (Kapitel == 4) { return  B_ReadyRune(slf,ItRu_Deathbolt);       };   // 125
            /*Kap 5+*/	        return  B_ReadyRune(slf,ItRu_Deathball);            // 165
        };
        
        if(slf.aivar[AIV_SelectSpell] == 10)
        {
            slf.aivar[AIV_SelectSpell] = 11;		// damit er nicht nochmal hier reinkommt
            
            dK_Mega = Hlp_Random(100);
//          if(!Wld_DetectNpcEx(slf,-1,NOFUNC,GIL_DMT,TRUE)) - не работает?
            if(dK_Mega <= 2)    { return  B_ReadyRune(slf,ItRu_Firerain); };        // dmg 500 шанс 3%
            if(dK_Mega <= 5)    { return  B_ReadyRune(slf,ItRu_Thunderstorm); };    // dmg 250 шанс 3% 
            if(dK_Mega <= 10)   { return  B_ReadyRune(slf,ItRu_LightningFlash); };  // dmg 150 шанс 5%
            /*11-99*/             return  B_ReadyRune(slf,ItRu_Firestorm);          // dmg 100 шанс 89%
        };
        
        return TRUE;		// hier immer TRUE zurьckgeben, ansonsten wird in B_SelectWeapon die Waffe wieder weggesteckt
    };
```

**Reply by Kor-Angar** [reply]
добавил комменты из MDK

**Reply by dosinabox** [reply]
```cpp
if (slf.guild == GIL_DMT) 
&& (slf.aivar[AIV_MagicUser] == MAGIC_ALWAYS) // излишняя проверка?
```

Да, потому что ищущих без этого аивара в игре нет.

**Reply by Kor-Angar** [reply]
обновил код [в ](https://github.com/dosinabox/g2nr_unofficial_update/discussions/520#discussioncomment-6521050)

**Reply by Kor-Angar** [reply]
можно наверное и так
```D
        if (slf.id == 1299)
        //if(C_IsNpc(slf,DMT_1299_OberDementor_DI))
        {
            //B_ReadySpell(slf,SPL_Skull,SPL_Cost_Skull);
            return  B_ReadyRune(slf,ItRu_Skull);
        };
```

**Reply by Kor-Angar** [reply]
думаю проверку на супербоса можно вынести и поставить раньше, проверки на ищущих.
```D
    if (slf.id == 1299)
    //if(C_IsNpc(slf,DMT_1299_OberDementor_DI))
    {
        //B_ReadySpell(slf,SPL_Skull,SPL_Cost_Skull);
        return  B_ReadyRune(slf,ItRu_Skull);
    };    
    
    // ------ Dementor -------
    if (slf.guild == GIL_DMT) 
    //&& (slf.aivar[AIV_MagicUser] == MAGIC_ALWAYS) // излишняя проверка?
    {
```

**Reply by dosinabox** [reply]
Мне такие проверки по id не нравятся. Это перестает быть читабельным. Число 1299 мне ни о чем не говорит, нужно идти в поиск. Лучше хотя бы добавить константу:
```cpp
const int DEMENTOR_BOSS = 1299;

id = DEMENTOR_BOSS;

if(slf.id == DEMENTOR_BOSS)
```

**Reply by Kor-Angar** [reply]
можно написать комментарий что речь идёт про DMT_1299_OberDementor_DI.  я напомню что эти проверки я стал использовать в связи с тем что при инъекции   могут поменяться InstanceID и проверка на  Hlp_GetInstanceID не будет срабатывать.

**Comment by dosinabox** [comment]
Фактически это означает, что ищущие смогут использовать заклинания массового поражения сразу, а не только после обычных, потому что начнет выполняться условие `if(slf.aivar[AIV_SelectSpell] == 10)`. Вообще логика тут странная и непредсказуемая.
```cpp
if(slf.aivar[AIV_SelectSpell] <= 0)
{
	dK_rnd = Hlp_Random(10); //0-9
	slf.aivar[AIV_SelectSpell] += dK_rnd; //возможен 0+0
};
if(slf.aivar[AIV_SelectSpell] < 10) //первым всегда выбирается обычный спелл, до 10 не хватает
{
	if(Kapitel <= 3)
	{
		B_ReadySpell(slf,SPL_InstantFireball,SPL_COST_InstantFireball);
	}
	else if(Kapitel == 4)
	{
		B_ReadySpell(slf,SPL_Deathbolt,SPL_COST_Deathbolt);
	}
	else
	{
		B_ReadySpell(slf,SPL_Deathball,SPL_COST_Deathball);
	};
}
else if(slf.aivar[AIV_SelectSpell] == 10) //10 возможно как минимум со второй попытки
{
	slf.aivar[AIV_SelectSpell] = 11; //условие с 11 никак не проверяется - тупят здесь?
	dK_Mega = Hlp_Random(100);
	if(dK_Mega <= 2)
	{
		B_ReadySpell(slf,SPL_Firerain,SPL_Cost_Firerain);
	}
	else if(dK_Mega <= 5)
	{
		B_ReadySpell(slf,SPL_Thunderstorm,SPL_Cost_Thunderstorm);
	}
	else if(dK_Mega <= 10)
	{
		B_ReadySpell(slf,SPL_LightningFlash,SPL_Cost_LightningFlash);
	}
	else
	{
		B_ReadySpell(slf,SPL_Firestorm,SPL_COST_Firestorm);
	};
}
else if(slf.aivar[AIV_SelectSpell] == 12) //обнуляется только при 12, что сейчас невозможно (?)
{
	slf.aivar[AIV_SelectSpell] = 0;
};
```

**Reply by Kor-Angar** [reply]
slf.aivar[AIV_SelectSpell] = 11;		// damit er nicht nochmal hier reinkommt

> //обнуляется только при 12, что сейчас невозможно (?)


как только AIV_SelectSpell принимает значение 10,  AIV_SelectSpell  переписывается на 11 и ищущий выбирает супер заклинание.

после того как ищущий сделает  каст  AIV_SelectSpell   увеличивается дл 12.

**Reply by dosinabox** [reply]
> как только AIV_SelectSpell принимает значение 10

Для этого dK_rnd должна выбросить ровно 10 за минимум две попытки. Если, например, рандом выдаст 5+6 или 4+4+4, то AIV_SelectSpell станет больше 12 и подходящих условий не останется, что приведет к зависанию ищущего (?).

> после того как ищущий сделает каст AIV_SelectSpell увеличивается до 12.

В Spell_Cast_*?

**Reply by Kor-Angar** [reply]
> Для этого dK_rnd должна выбросить ровно 10 за минимум две попытки. Если, например, рандом выдаст 5+6 или 4+4+4, то AIV_SelectSpell станет больше 12 и подходящих условий не останется, что приведет к зависанию ищущего (?).

нет.

происходит рандом. принимается значение 0-9.  для упрощение лучше если значение будет 1-9, тогда рандом подряд не вызовется при 0.

после того как рандом принял значение ищущий достатёт фаербол. каждый каст будет увеличивать AIV_SelectSpell  на 1. 

то есть если выпално, например 5, то ищущий сделает 5 кастов фаерболом, после чего AIV_SelectSpell   примет значение 10 и ищущий **сменить** руну на более мощную. AIV_SelectSpell    = 11.

мощной руной ищущий делает один(!) каст и AIV_SelectSpell   принимает значение 12, после чего AIV_SelectSpell   обнуляется.

**Reply by dosinabox** [reply]
> каждый каст будет увеличивать AIV_SelectSpell на 1.

Точно. Забываю про Spell_Cast_*.

В таком случае, твое предлагаемое изменение `slf.aivar[AIV_SelectSpell] = dK_rnd+1; // 1-10` позволит ищущим сразу получить 10 в AIV_SelectSpell и скастовать огненный дождь при первом виде ГГ. Что не очень хорошо.

**Reply by Kor-Angar** [reply]
да. я это уже понял. и выше написал. думаю вот так попробовать

```D
            //dK_rnd = Hlp_Random(10);
            //slf.aivar[AIV_SelectSpell] += dK_rnd; //+ 0-9
            dK_rnd = Hlp_Random(9);
            slf.aivar[AIV_SelectSpell] = dK_rnd+1; // 1-9
```

**Reply by dosinabox** [reply]
```cpp
slf.aivar[AIV_SelectSpell] = Hlp_Random(9) + 1;
```

**Reply by Kor-Angar** [reply]
угу

**Reply by Kor-Angar** [reply]
slf.aivar[AIV_SelectSpell] = Hlp_Random(9) + 1; // 1-9

**Comment by Kor-Angar** [comment]
смотри какой косяк нашёл
![image](https://github.com/dosinabox/g2nr_unofficial_update/assets/49514764/c87bbb37-83f9-4ac8-8b00-09ccb2f062b2)


в 4 главе малый огненный шторм слабее чем **ItRu_Deathbolt** ;-)

**Reply by dosinabox** [reply]
Предлагаешь заменить ItRu_Firestorm на ItRu_Deathbolt? Или проверять главу и для 11-99?

**Reply by Kor-Angar** [reply]
надо подумать. может быть в 4 главе давать вместо ItRu_Firestorm(100 ущерба)  Feuersturm(150 ущерба)
а что в 5 главе давать я хз :-)

**Reply by Kor-Angar** [reply]
тут просто табличку надо сделать. со столбиком по главам. например







  | 3 глава | 4 глава | 5 глава | 6 глава
-- | -- | -- | -- | --
база. AIV_SelectSpell < 10 | 75 | 125 | 165 | 165
шанс 89% | 100 | 150 | ? | ?
шанс 5% | 150 | ? | ? | ?
шанс 3% | 250 | 250 | 250 | 250
шанс 3% | 500 | 500 | 500 | 500

**Reply by dosinabox** [reply]
Deathball (165) подходит под все вопросительные знаки

**Reply by Kor-Angar** [reply]
неа. тогда у тебя сама логика ИИ ломается. ведь при AIV_SelectSpell  == 10 Ищущий должен делать "супер удар" ;-)

**Reply by Kor-Angar** [reply]
можно вместо молнии на поздних главах дать ему ItRu_IceCube или волну льда. но лучше эту тему поднять на форуме. пусть коллективный разум решает что имба что нет, что дисбаланс что нет.

**Reply by dosinabox** [reply]
Пока так оставлю:
```cpp
if(dK_Mega <= 2)
{
	B_ReadySpell(slf,SPL_Firerain,SPL_Cost_Firerain);
}
else if(dK_Mega <= 5)
{
	B_ReadySpell(slf,SPL_Thunderstorm,SPL_Cost_Thunderstorm);
}
else if(dK_Mega <= 10)
{
	B_ReadySpell(slf,SPL_LightningFlash,SPL_Cost_LightningFlash);
}
else
{
	if(Kapitel <= 3)
	{
		B_ReadySpell(slf,SPL_Firestorm,SPL_COST_Firestorm);
	}
	else
	{
		//TODO нужно что-то более мощное
		B_ReadySpell(slf,SPL_Deathball,SPL_COST_Deathball);
	};
};
```

**Reply by Kor-Angar** [reply]
молния для 5 и 6 главы тоже слабая

**Comment by Kor-Angar** [comment]
где то писал ранее. кажется на форуме. 

```D
func int B_ReadyRune(var C_NPC slf, var int rune)
{
    
    if (!Npc_GetInvItem(slf, rune))
    {
        CreateInvItem(slf, rune);
    };
    Hlp_PrintConsole(Str_Format("  B_ReadyRune[%s.%i]  --> rune %s", slf.name, slf.id,item.description));
    
    var int SpellId;    SpellId = item.spell;
    //var int SpellMana;  SpellMana = item.COUNT[1];
    
    //B_ReadySpell(slf, SpellId, SpellMana);
    B_ReadySpell(slf, SpellId, 0);
    return TRUE;   
};
```

**Comment by Kor-Angar** [comment]
**GIL_PAL**

**Reply by Kor-Angar** [reply]
это новодел?
```D
        if((Npc_GetDistToNpc(slf,oth) > FIGHT_DIST_MELEE) && C_NpcIsEvil(oth))
        {
            if(oth.protection[PROT_MAGIC] > SPL_Damage_PalHolyBolt)
            {
                B_ReadySpell(slf,SPL_PalRepelEvil,SPL_Cost_PalRepelEvil);
            }
            else
            {
                B_ReadySpell(slf,SPL_PalHolyBolt,SPL_Cost_PalHolyBolt);
            };
            return TRUE;
        };
```
тут было бы логично делать проверку на уровень паладина. если паладин 50 уровня и выше давать сильную руну. если паладин 40 уровня давать HolyBolt. если ущерб руны ниже защиты от магии возвращать false.

**Reply by Kor-Angar** [reply]
при атаке на замок орками у паладинов нет шанса если они атакуют орков рунами. соответственно в минентале надо или им запрещать использовать руну или давать более сильную руну.

при этом замечу что атака орков в готики 2 классик и готики 2 нв из-за правок баланса две большие разница. в классике паладины достаточно легко отбивают атаку ;-)

**Comment by Kor-Angar** [comment]
скелет маг

**Reply by Kor-Angar** [reply]
похоже что скелет маг  использует ItRu_IceCube всего один раз?

```D
    // ------ Skelett Magier ------
    if(slf.guild == GIL_SKELETON_MAGE)
    {
        //if(!Npc_HasItems(slf,ItRu_SumSkel))        { CreateInvItems(slf,ItRu_SumSkel,1); };
        //if(!Npc_HasItems(slf,ItRu_IceCube))        { CreateInvItems(slf,ItRu_IceCube,1); };
        //if(!Npc_HasItems(slf,ItRu_Icebolt))        { CreateInvItems(slf,ItRu_Icebolt,1); };
        
        // ------ Spruchzyklus bei SUMMON beginnen ------
        if(slf.aivar[AIV_SelectSpell] >= 6)
        {
            slf.aivar[AIV_SelectSpell] = 1;
        };
        
        if(!Npc_IsInState(oth,ZS_MagicFreeze) && (slf.aivar[AIV_SelectSpell] == 0))
        {
            //B_ReadySpell(slf,SPL_IceCube,SPL_Cost_IceCube);
            //return TRUE;
            return  B_ReadyRune(slf,ItRu_IceCube);
        };
        
        if(slf.aivar[AIV_SelectSpell] == 1)
        {
            //B_ReadySpell(slf,SPL_SummonSkeleton,SPL_Cost_SummonSkeleton);
            //return TRUE;
            return  B_ReadyRune(slf,ItRu_SumSkel);
        };
        
        //B_ReadySpell(slf,SPL_Icebolt,SPL_COST_Icebolt);
        //return TRUE;
        return  B_ReadyRune(slf,ItRu_Icebolt);
    };
```

**Reply by Kor-Angar** [reply]
похоже что скелет маг  использует ItRu_IceCube всего один раз?

**Comment by dosinabox** [comment]
обновлено:
https://github.com/dosinabox/g2nr_unofficial_update/pull/523/files#diff-b8dc8e49444d13bfb7bb26d8db485d34700ba7502a06bfcd868c20d51e3a9682

https://github.com/dosinabox/g2nr_unofficial_update/blob/work/PrjGOTHIC/AI/Magic/B_SelectSpell.d

**Comment by Kor-Angar** [comment]
в начале True или  false должно было быть. в мдк  было а у тебя пропало. то есть друзья героя магией не убивают, а усыпляют. 

```D
func int B_SelectSpell(var C_Npc slf,var C_Npc oth)
{
	var int dK_Mega;
	if((slf.npcType == NPCTYPE_FRIEND) && Npc_IsPlayer(oth) && (oth.guild < GIL_SEPERATOR_HUM))
	{
		if((slf.guild == GIL_KDF) || (slf.guild == GIL_KDW) || (slf.aivar[AIV_MagicUser] == MAGIC_ALWAYS))
		{
			if(!Npc_HasItems(slf,ItRu_Sleep))
			{
				CreateInvItems(slf,ItRu_Sleep,1);
			};
			B_ReadySpell(slf,SPL_Sleep,SPL_Cost_Sleep);
			return TRUE;
		};
                return false;
	};

```

**Comment by Kor-Angar** [comment]
создание руны лучше засунуть непосредственно перед вызовом спелла, так как эти руны всё равно потом придётся скриптами удалять. непонятно зачем создавать руны которые не будут использованы :-D

**Reply by Kor-Angar** [reply]
```D
		if (C_IsNpc(slf,DMT_1299_OberDementor_DI))
		{
			if(!Npc_HasItems(slf,ItRu_Skull))
			{
				CreateInvItems(slf,ItRu_Skull,1);
			};
			B_ReadySpell(slf,SPL_Skull,SPL_Cost_Skull);
			return TRUE;
		};
```
